### PR TITLE
observer extension: Add generic endpoint watcher

### PR DESCRIPTION
**Pull request by asuresh4**
**Description:** Add `EndpointsWatcher`,  a struct with utilities required to poll and track endpoints in a generic manner. It is an implementation of `observer.Observable` and will help in abstracting some logic that's required by poll based observers. This struct is meant to be embedded by observers that list `observer.Endpoint`. Will also open a separate PR with changes that make use of this update.

The diff also has some minor refactoring.

**Testing:** Added unit tests.

**Documentation:** Added inline comments.

### CODE REVIEW

File: extension/observer/endpointswatcher.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+type EndpointsWatcher struct {
+	ListEndpoints     func() []Endpoint
+	RefreshInterval   time.Duration
+	existingEndpoints map[string]Endpoint
+	stop              chan struct{}
+}
+
+// ListAndWatch runs ListEndpoints on a regular interval and keeps the list.
+func (ew *EndpointsWatcher) ListAndWatch(listener Notify) {
+	ew.existingEndpoints = make(map[string]Endpoint)
```
Comment (jrcamp) [inline extension/observer/endpointswatcher.go:35]:
```suggestion
	ew.existingEndpoints = map[string]Endpoint{}
```

File: extension/observer/endpointswatcher.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+// EndpointsWatcher provides a generic mechanism to run ListEndpoints every
+// RefreshInterval and report any new or removed endpoints using Notify
+// passed into ListAndWatch. Any observer that lists endpoints can make
+// use of EndpointsWatcher to poll for endpoints by embedding this struct
+// in the observer struct.
+type EndpointsWatcher struct {
+	ListEndpoints     func() []Endpoint
+	RefreshInterval   time.Duration
+	existingEndpoints map[string]Endpoint
```
Comment (jrcamp) [inline extension/observer/endpointswatcher.go:29]: Might be worth adding a typedef for ID.

File: extension/observer/endpointswatcher.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	ew.existingEndpoints = make(map[string]Endpoint)
+	ew.stop = make(chan struct{})
+
+	ticker := time.NewTicker(ew.RefreshInterval)
+
+	// Do the initial listing immediately so that services can be monitored ASAP.
+	ew.refreshEndpoints(listener)
+
+	go func() {
+		for {
```
Comment (jrcamp) [inline extension/observer/endpointswatcher.go:44]: Would move ticker into this function so you can keep it together with the stop:

```go
ticker := time.NewTicker(ew.RefreshInterval)
defer ticker.Stop()
```

File: extension/observer/endpointswatcher.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+// refreshEndpoints updates the listener with the latest list
+// of active endpoints.
+func (ew *EndpointsWatcher) refreshEndpoints(listener Notify) {
+	latestEndpoints := ew.ListEndpoints()
+
+	// Create map from ID to endpoint for lookup.
+	latestEndpointsMap := make(map[string]Endpoint, len(latestEndpoints))
+	for _, e := range latestEndpoints {
+		latestEndpointsMap[e.ID] = e
```
Comment (jrcamp) [inline extension/observer/endpointswatcher.go:64]: Doesn't look like it's using the value below so maybe just make this a map[string]bool?

File: extension/observer/endpointswatcher.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+	// If endpoint present in existingEndpoints does not exist in the latest
+	// list, it needs to be removed.
+	for id, e := range ew.existingEndpoints {
+		if _, ok := latestEndpointsMap[e.ID]; !ok {
+			delete(ew.existingEndpoints, id)
+			removedEndpoints = append(removedEndpoints, e)
+		}
+	}
+
```
Comment (jrcamp) [inline extension/observer/endpointswatcher.go:85]: What about changed endpoints?

File: extension/observer/endpoints.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+		}, nil
+	case Port:
+		ruleTypes["port"] = true
+		return map[string]interface{}{
+			"type":     ruleTypes,
+			"endpoint": endpoint.Target,
+			"name":     o.Name,
+			"port":     o.Port,
+			"pod": map[string]interface{}{
+				"name":   o.Pod.Name,
```
Comment (jrcamp) [inline extension/observer/endpoints.go:78]: Just noticed I forgot to add annotations here like in the pod above 🤦. Do you mind adding it?

Comment (asuresh4) [reply extension/observer/endpoints.go:78]: Yep, I'll add it.

File: extension/observer/endpointswatcher.go (hunk 5 above)
Comment (asuresh4) [reply extension/observer/endpointswatcher.go:85]: Good catch will add.

File: extension/observer/endpointswatcher.go (hunk 2 above)
Comment (asuresh4) [reply extension/observer/endpointswatcher.go:29]: 👍 sounds good, will do.

File: extension/observer/endpointswatcher.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+				ew.existingEndpoints[e.ID] = e
+				updatedEndpoints = append(updatedEndpoints, e)
+			}
+		}
+	}
+
+	// If endpoint present in existingEndpoints does not exist in the latest
+	// list, it needs to be removed.
+	for id, e := range ew.existingEndpoints {
+		if _, ok := latestEndpointsMap[e.ID]; !ok {
```
Comment (jrcamp) [inline extension/observer/endpointswatcher.go:87]: Since this is a bool map now can just do:

```go
if latestEndpointsMap[e.ID] {
...
}
```

Comment (asuresh4) [reply extension/observer/endpointswatcher.go]: Updated.

**Review by jrcamp — APPROVED** [review]

**Pull request merged**